In [1]:
import pandas as pd

bitcoin = pd.read_csv("/Users/micheal/Downloads/bitcoin_data.csv")

bitcoin.head()

,trans_date,price_usd,code_size,sent_by_address,transactions,mining_profitability,sent_in_usd,transaction_fees,median_transaction_fee,confirmation_time,market_cap,transaction_value,median_transaction_value,tweets,google_trends,fee_to_reward,active_addresses,top_100_cap
0,2014-04-09,447.110,243742,122401,64992,38.678,239212041,0.1390,0.0447,8.944,5587996110,4115,100.420,8193,11.193,0.502,179274,19.211
1,2014-04-10,407.703,254334,124298,67872,31.917,310367120,0.0864,0.0408,9.290,5020359905,4876,120.946,15039,13.854,0.371,182996,19.166
2,2014-04-11,403.328,223656,117926,65563,36.814,292756343,0.0858,0.0403,8.521,5068919648,4819,108.481,14907,12.997,0.330,169622,19.141
3,2014-04-12,425.310,169964,87828,47579,34.498,157264741,0.0811,0.0425,8.944,5329088351,3574,84.408,7582,9.229,0.226,132283,19.145
4,2014-04-13,415.876,162260,94828,50981,34.141,154756367,0.0840,0.0416,7.701,5150414718,3332,69.681,10674,10.385,0.221,172708,19.131


In [2]:
bitcoin.shape

(3180, 18)

In [ ]:
bitcoin.shape

(3180, 18)

In [3]:
bitcoin.columns

Index(['trans_date', 'price_usd', 'code_size', 'sent_by_address',
       'transactions', 'mining_profitability', 'sent_in_usd',
       'transaction_fees', 'median_transaction_fee', 'confirmation_time',
       'market_cap', 'transaction_value', 'median_transaction_value', 'tweets',
       'google_trends', 'fee_to_reward', 'active_addresses', 'top_100_cap'],
      dtype='str')

In [4]:
bitcoin.info()

<class 'pandas.DataFrame'>
RangeIndex: 3180 entries, 0 to 3179
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   trans_date                3180 non-null   str    
 1   price_usd                 3180 non-null   float64
 2   code_size                 3180 non-null   int64  
 3   sent_by_address           3180 non-null   int64  
 4   transactions              3180 non-null   int64  
 5   mining_profitability      3180 non-null   float64
 6   sent_in_usd               3180 non-null   int64  
 7   transaction_fees          3180 non-null   float64
 8   median_transaction_fee    3180 non-null   float64
 9   confirmation_time         3180 non-null   float64
 10  market_cap                3180 non-null   int64  
 11  transaction_value         3180 non-null   int64  
 12  median_transaction_value  3180 non-null   float64
 13  tweets                    3180 non-null   int64  
 14  google_trends      

## Check Missing value

In [6]:
bitcoin.isnull().sum()

trans_date                  0
price_usd                   0
code_size                   0
sent_by_address             0
transactions                0
mining_profitability        0
sent_in_usd                 0
transaction_fees            0
median_transaction_fee      0
confirmation_time           0
market_cap                  0
transaction_value           0
median_transaction_value    0
tweets                      0
google_trends               0
fee_to_reward               0
active_addresses            0
top_100_cap                 0
dtype: int64

In [7]:
bitcoin.info()

<class 'pandas.DataFrame'>
RangeIndex: 3180 entries, 0 to 3179
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   trans_date                3180 non-null   str    
 1   price_usd                 3180 non-null   float64
 2   code_size                 3180 non-null   int64  
 3   sent_by_address           3180 non-null   int64  
 4   transactions              3180 non-null   int64  
 5   mining_profitability      3180 non-null   float64
 6   sent_in_usd               3180 non-null   int64  
 7   transaction_fees          3180 non-null   float64
 8   median_transaction_fee    3180 non-null   float64
 9   confirmation_time         3180 non-null   float64
 10  market_cap                3180 non-null   int64  
 11  transaction_value         3180 non-null   int64  
 12  median_transaction_value  3180 non-null   float64
 13  tweets                    3180 non-null   int64  
 14  google_trends      

In [8]:
bitcoin.isnull().sum()

trans_date                  0
price_usd                   0
code_size                   0
sent_by_address             0
transactions                0
mining_profitability        0
sent_in_usd                 0
transaction_fees            0
median_transaction_fee      0
confirmation_time           0
market_cap                  0
transaction_value           0
median_transaction_value    0
tweets                      0
google_trends               0
fee_to_reward               0
active_addresses            0
top_100_cap                 0
dtype: int64

In [9]:
bitcoin.dtypes

trans_date                      str
price_usd                   float64
code_size                     int64
sent_by_address               int64
transactions                  int64
mining_profitability        float64
sent_in_usd                   int64
transaction_fees            float64
median_transaction_fee      float64
confirmation_time           float64
market_cap                    int64
transaction_value             int64
median_transaction_value    float64
tweets                        int64
google_trends               float64
fee_to_reward               float64
active_addresses              int64
top_100_cap                 float64
dtype: object

In [10]:
bitcoin["trans_date"] = pd.to_datetime(bitcoin["trans_date"])

In [11]:
bitcoin["trans_date"].dtype

dtype('<M8[us]')

In [12]:
bitcoin.duplicated().sum()

np.int64(0)

In [13]:
bitcoin.describe()

,trans_date,price_usd,code_size,sent_by_address,transactions,mining_profitability,sent_in_usd,transaction_fees,median_transaction_fee,confirmation_time,market_cap,transaction_value,median_transaction_value,tweets,google_trends,fee_to_reward,active_addresses,top_100_cap
count,3180,3180.000000,3180.000000,3180.000000,3180.000000,3180.000000,3.180000e+03,3180.000000,3180.000000,3180.000000,3.180000e+03,3.180000e+03,3180.000000,3180.000000,3180.000000,3180.000000,3.180000e+03,3180.000000
mean,2018-08-15 12:00:00,12210.656699,702125.407547,370935.080818,236883.029560,1.957467,1.192082e+10,2.704139,1.309816,9.787805,2.257468e+11,1.246725e+05,406.650023,52725.178616,51.196511,4.113218,5.769545e+05,17.179965
min,2014-04-09 00:00:00,197.577000,162260.000000,70902.000000,47579.000000,0.052500,9.471432e+07,0.024100,0.009970,0.000000,2.655832e+09,1.166000e+03,0.091700,0.000000,5.268000,0.189000,0.000000e+00,0.000000
25%,2016-06-11 18:00:00,603.998000,583895.250000,288086.750000,192269.500000,0.174000,9.354786e+08,0.154750,0.062175,9.057000,9.040045e+09,5.850500e+03,129.237750,21911.750000,9.880500,0.931000,4.152070e+05,15.153500
50%,2018-08-15 12:00:00,6571.500000,740537.000000,390244.500000,252094.500000,0.377500,4.374385e+09,0.887500,0.304500,9.664000,1.135541e+11,2.560200e+04,305.525500,30191.500000,35.718500,1.608000,5.956855e+05,17.223500
75%,2020-10-18 06:00:00,15238.500000,879222.000000,476081.250000,296568.000000,1.467500,1.152109e+10,2.380500,0.938500,10.360000,2.580889e+11,1.000435e+05,542.380750,75057.500000,63.694250,5.380500,7.662425e+05,19.207500
max,2022-12-22 00:00:00,67547.000000,998092.000000,730199.000000,490459.000000,40.706000,3.833295e+11,62.779000,34.095000,24.828000,1.274513e+12,2.450460e+06,5417.000000,363566.000000,616.867000,43.580000,1.196069e+06,21.809000
std,NaN,16007.272397,208326.491540,132993.425162,84283.742917,5.041295,2.244176e+10,5.883913,3.093767,1.183435,3.026721e+11,2.371334e+05,452.258229,47209.368812,57.647266,5.351474,2.405334e+05,2.401756


# Part 1: Bitcoin Data

## Query 1 — Difficulty

Calculate the ratio of `code_size` to `transactions` for each date.

The calculated column should be named `difficulty`, and all decimal places resulting from the division should be retained.

In [17]:
bitcoin["difficulty"] = bitcoin["code_size"] / bitcoin["transactions"]

query1 = bitcoin[["trans_date", "difficulty"]]

query1

,trans_date,difficulty
0,2014-04-09,3.750339
1,2014-04-10,3.747260
2,2014-04-11,3.411314
3,2014-04-12,3.572248
4,2014-04-13,3.182754
...,...,...
3175,2022-12-18,2.406586
3176,2022-12-19,2.726718
3177,2022-12-20,2.476501
3178,2022-12-21,2.398945


2.Write a query that returns the trans_date and the product of median_transaction_fee
and transactions for days where the median_transaction_fee is more than 0.5. Label
the calculated column daily_cost.

## Query 2 — Daily Transaction Cost

Return the `trans_date` and the product of `median_transaction_fee` and `transactions` for days where `median_transaction_fee` is greater than 0.5.

The calculated column should be named `daily_cost`.

In [18]:
query2 = bitcoin.loc[
    bitcoin["median_transaction_fee"] > 0.5,
    ["trans_date", "median_transaction_fee", "transactions"]
].copy()

query2["daily_cost"] = (
    query2["median_transaction_fee"] * query2["transactions"]
)

query2 = query2[["trans_date", "daily_cost"]]

query2

,trans_date,daily_cost
1055,2017-02-27,147893.720
1056,2017-02-28,173689.374
1057,2017-03-01,171762.016
1058,2017-03-02,184391.760
1059,2017-03-03,212747.329
...,...,...
3159,2022-12-02,233847.600
3160,2022-12-03,191877.300
3161,2022-12-04,139929.790
3162,2022-12-05,173664.012


### Interpretation

The query returned 1,238 observations where the median transaction fee was greater than 0.5. The `daily_cost` was calculated by multiplying the median transaction fee by the number of transactions for each qualifying day. For example, the daily cost on February 27, 2017 was 147,893.720.

3.Write a query that returns the trans_date and the sent_in_usd divided by the number of
transactions on days where there were more than 400,000 transactions. Show all
expected decimals. Name the output column avg_transaction.

In [19]:
query3 = bitcoin.loc[
    bitcoin["transactions"] > 400000,
    ["trans_date", "sent_in_usd", "transactions"]
].copy()

query3["avg_transaction"] = (
    query3["sent_in_usd"] / query3["transactions"]
)

query3 = query3[["trans_date", "avg_transaction"]]

query3

,trans_date,avg_transaction
1332,2017-12-01,47832.586022
1335,2017-12-04,63679.190437
1336,2017-12-05,69682.831110
1337,2017-12-06,80823.462089
1340,2017-12-09,90671.727289
1344,2017-12-13,109426.002587
1345,2017-12-14,77685.261084
1346,2017-12-15,94321.307171
1366,2018-01-04,51624.933436
1819,2019-04-02,17293.121326


Write a query that returns the average price_usd (rounded to 2 decimal places, output
as MONEY datatype) and the total amount sent_in_usd (output as MONEY datatype)
for all transactions that occurred in May of 2015. Name the output columns avg_price
and total_transaction_amount, respectively.

In [20]:
query4 = bitcoin.loc[
    (bitcoin["trans_date"] >= "2015-05-01") &
    (bitcoin["trans_date"] < "2015-06-01")
]

query4_result = pd.DataFrame({
    "avg_price": [query4["price_usd"].mean()],
    "total_transaction_amount": [query4["sent_in_usd"].sum()]
})

query4_result["avg_price"] = query4_result["avg_price"].round(2)

query4_result

,avg_price,total_transaction_amount
0,237.3,6584198028


### Interpretation

For May 2015, the average Bitcoin price was $237.30. The total amount sent during the month was $6,584,198,028. The average price was rounded to two decimal places as required, while the total transaction amount represents the sum of `sent_in_usd` across all observations in May 2015.

5.Write a query that returns the smallest and largest market_cap values as whole
numbers and the lowest and highest price_usd rounded to 2 decimals. Name these
columns as low_cap, high_cap, low_price_usd, and high_price_usd respectively.

In [21]:
query5 = pd.DataFrame({
    "low_cap": [bitcoin["market_cap"].min()],
    "high_cap": [bitcoin["market_cap"].max()],
    "low_price_usd": [bitcoin["price_usd"].min()],
    "high_price_usd": [bitcoin["price_usd"].max()]
})

query5["low_price_usd"] = query5["low_price_usd"].round(2)
query5["high_price_usd"] = query5["high_price_usd"].round(2)

query5

,low_cap,high_cap,low_price_usd,high_price_usd
0,2655832477,1274512840508,197.58,67547.0


### Interpretation

6.Across the full Bitcoin dataset, the lowest recorded market capitalization was 2,655,832,477, while the highest was 1,274,512,840,508. The Bitcoin price ranged from a minimum of $197.58 to a maximum of $67,547.00. These results show substantial variation in both Bitcoin's market capitalization and price over the period covered by the dataset.

In [23]:
query6_data = bitcoin.loc[
    (bitcoin["trans_date"] >= "2017-08-10") &
    (bitcoin["trans_date"] <= "2019-08-10")
]

query6 = pd.DataFrame({
    "lowest_fee": [query6_data["median_transaction_fee"].min()],
    "highest_fee": [query6_data["median_transaction_fee"].max()],
    "avg_fee": [query6_data["transaction_fees"].mean()]
})

query6["avg_fee"] = query6["avg_fee"].round(5)

query6

,lowest_fee,highest_fee,avg_fee
0,0.0254,34.095,3.84405


7.Write a query that determines the average number of transactions and average
sent_by_address for days with more than 350,000 transactions. Name these columns
avg_transactions and avg_sent_addresses respectively.

In [24]:
query7_data = bitcoin.loc[
    bitcoin["transactions"] > 350000
]

query7 = pd.DataFrame({
    "avg_transactions": [query7_data["transactions"].mean()],
    "avg_sent_addresses": [query7_data["sent_by_address"].mean()]
})

query7

,avg_transactions,avg_sent_addresses
0,373760.788571,506180.691429


8.Write a query that shows the average google_trends score (as a whole number) and
the total number of tweets. Name the columns avg_google_trends and total_tweets.

In [25]:
query8 = pd.DataFrame({
    "avg_google_trends": [bitcoin["google_trends"].mean()],
    "total_tweets": [bitcoin["tweets"].sum()]
})

query8["avg_google_trends"] = query8["avg_google_trends"].round().astype(int)

query8

,avg_google_trends,total_tweets
0,51,167666068


9.Write a query that returns the minimum, maximum, and average confirmation_time for
all transactions. Round the average to three decimal places. Name the output columns
min_confirmation_time, max_confirmation_time, and avg_confirmation_time.

In [26]:
query9 = pd.DataFrame({
    "min_confirmation_time": [bitcoin["confirmation_time"].min()],
    "max_confirmation_time": [bitcoin["confirmation_time"].max()],
    "avg_confirmation_time": [bitcoin["confirmation_time"].mean()]
})

query9["avg_confirmation_time"] = query9["avg_confirmation_time"].round(3)

query9

,min_confirmation_time,max_confirmation_time,avg_confirmation_time
0,0.0,24.828,9.788


10.Write a query that returns the averages for the following columns (all rounded to 2
decimals): price_usd, mining_profitability, and transaction_fees for all transactions
that happened between March 1, 2020, and November 30, 2020. Name the columns:
avg_price_usd, avg_mining_profitability, and avg_transaction_fees respectively.

In [27]:
query10_data = bitcoin.loc[
    (bitcoin["trans_date"] >= "2020-03-01") &
    (bitcoin["trans_date"] <= "2020-11-30")
]

query10 = pd.DataFrame({
    "avg_price_usd": [query10_data["price_usd"].mean()],
    "avg_mining_profitability": [
        query10_data["mining_profitability"].mean()
    ],
    "avg_transaction_fees": [
        query10_data["transaction_fees"].mean()
    ]
})

query10 = query10.round(2)

query10

,avg_price_usd,avg_mining_profitability,avg_transaction_fees
0,10329.18,0.1,2.9


In [28]:
customer_spending = pd.read_csv(
    "/Users/micheal/Downloads/customer_spending.csv"
)

customer_spending.head()

,sale_date,sale_year,sale_month,age,gender,country,state,category,sub_category,quantity,unit_cost,unit_price,cost,revenue
0,2016-02-19,2016,February,29,F,United States,Washington,Accessories,Tires and Tubes,1,80.00,109.000000,80,109
1,2016-02-20,2016,February,29,F,United States,Washington,Clothing,Gloves,2,24.50,28.500000,49,57
2,2016-02-27,2016,February,29,F,United States,Washington,Accessories,Tires and Tubes,3,3.67,5.000000,11,15
3,2016-03-12,2016,March,29,F,United States,Washington,Accessories,Tires and Tubes,2,87.50,116.500000,175,233
4,2016-03-12,2016,March,29,F,United States,Washington,Accessories,Tires and Tubes,3,35.00,41.666667,105,125


In [29]:
customer_spending.shape

(34866, 14)

In [30]:
customer_spending.isnull().sum()

sale_date       0
sale_year       0
sale_month      0
age             0
gender          0
country         0
state           0
category        0
sub_category    0
quantity        0
unit_cost       0
unit_price      0
cost            0
revenue         0
dtype: int64

In [31]:
customer_spending.dtypes

sale_date           str
sale_year         int64
sale_month          str
age               int64
gender              str
country             str
state               str
category            str
sub_category        str
quantity          int64
unit_cost       float64
unit_price      float64
cost              int64
revenue           int64
dtype: object

In [32]:
customer_spending.dtypes

sale_date           str
sale_year         int64
sale_month          str
age               int64
gender              str
country             str
state               str
category            str
sub_category        str
quantity          int64
unit_cost       float64
unit_price      float64
cost              int64
revenue           int64
dtype: object

1.Write a query that returns each category and the corresponding total revenue for that
category for the sale_year 2016. The aggregated column should be named
total_revenue. The output should be arranged alphabetically.

In [33]:
query1_customer = (
    customer_spending[customer_spending["sale_year"] == 2016]
    .groupby("category", as_index=False)["revenue"]
    .sum()
    .rename(columns={"revenue": "total_revenue"})
    .sort_values("category")
)

query1_customer

,category,total_revenue
0,Accessories,4594897
1,Bikes,5722257
2,Clothing,2079651


### Interpretation

In 2016, Bikes generated the highest total revenue among the three categories, with total revenue of 5,722,257. Accessories generated 4,594,897, while Clothing generated 2,079,651. The results show that revenue differed substantially across the product categories during 2016.

2.Write a query that returns a list of sub_categories and their corresponding average
unit_price (named avg_unit_price), average unit_cost named avg_unit_cost, as well as
the difference between these two values (named margin) for the sale_year 2015. Round
all values to two decimal places. Organize the results alphabetically.

In [34]:
query2_customer = (
    customer_spending[customer_spending["sale_year"] == 2015]
    .groupby("sub_category")
    .agg(
        avg_unit_price=("unit_price", "mean"),
        avg_unit_cost=("unit_cost", "mean")
    )
    .reset_index()
)

query2_customer["margin"] = (
    query2_customer["avg_unit_price"] -
    query2_customer["avg_unit_cost"]
)

query2_customer[
    ["avg_unit_price", "avg_unit_cost", "margin"]
] = query2_customer[
    ["avg_unit_price", "avg_unit_cost", "margin"]
].round(2)

query2_customer = query2_customer.sort_values("sub_category")

query2_customer

,sub_category,avg_unit_price,avg_unit_cost,margin
0,Bike Stands,602.00,528.61,73.40
1,Bottles and Cages,75.82,66.88,8.94
2,Caps,99.96,90.38,9.58
3,Cleaners,94.85,84.20,10.65
4,Helmets,368.32,322.19,46.13
5,Hydration Packs,545.69,481.44,64.25
6,Jerseys,511.79,462.61,49.18
7,Mountain Bikes,1108.03,1147.46,-39.43
8,Road Bikes,799.99,818.16,-18.17
9,Shorts,728.04,679.28,48.76


3.Write a query that returns the sale_year and corresponding total number of female
buyers (gender) for each sale_year who made purchases in the Clothing category.
Name the aggregated column total_female_buyers.

In [35]:
query3_customer = (
    customer_spending[
        (customer_spending["category"] == "Clothing") &
        (customer_spending["gender"] == "F")
    ]
    .groupby("sale_year")
    .size()
    .reset_index(name="total_female_buyers")
)

query3_customer

,sale_year,total_female_buyers
0,2015,1037
1,2016,1468


### Interpretation

Among the female customer purchase records in the Clothing category, 1,037 were recorded in 2015 and 1,468 were recorded in 2016. The number of female Clothing purchase records was therefore higher in 2016 than in 2015.

4.Write a query that returns the age, sub_cateogry, average quantity (as a whole
number, named avg_quantity), and average cost of products (rounded to 2 decimals,
named avg_cost) purchased by each age and sub_category. Organize the data by age,
oldest to youngest, and then by sub_category alphabetically.

In [36]:
query4_customer = (
    customer_spending
    .groupby(["age", "sub_category"])
    .agg(
        avg_quantity=("quantity", "mean"),
        avg_cost=("cost", "mean")
    )
    .reset_index()
)

query4_customer["avg_quantity"] = (
    query4_customer["avg_quantity"].round().astype(int)
)

query4_customer["avg_cost"] = (
    query4_customer["avg_cost"].round(2)
)

query4_customer = query4_customer.sort_values(
    ["age", "sub_category"],
    ascending=[False, True]
)

query4_customer

,age,sub_category,avg_quantity,avg_cost
889,87,Bike Racks,3,240.00
890,87,Tires and Tubes,2,44.50
886,86,Fenders,3,637.00
887,86,Tires and Tubes,2,166.50
888,86,Vests,1,1842.00
...,...,...,...,...
11,17,Shorts,2,787.50
12,17,Socks,2,171.00
13,17,Tires and Tubes,2,204.34
14,17,Touring Bikes,2,1278.71


### Interpretation

The results show how average purchase quantity and average product cost vary across customer ages and sub-categories. The dataset includes customers ranging from age 17 to age 87. For each age, the results are further divided by sub-category, allowing differences in purchasing quantity and product cost to be examined across both customer age and product type.

5.Write a query that returns a list of countries where more than 900 transactions were
made by customers between the ages of 18-25 (inclusive).

In [37]:
query5_customer = (
    customer_spending[
        customer_spending["age"].between(18, 25)
    ]
    .groupby("country")
    .size()
    .reset_index(name="total_transactions")
)

query5_customer = query5_customer[
    query5_customer["total_transactions"] > 900
]

query5_customer

,country,total_transactions
1,Germany,976
2,United Kingdom,1101
3,United States,2823


### Interpretation

Among customers aged 18 to 25, three countries recorded more than 900 transactions. The United States had the highest number of transactions with 2,823, followed by the United Kingdom with 1,101 and Germany with 976. These results indicate that the volume of transactions from customers aged 18–25 was substantially higher in the United States than in the other countries meeting the criterion.

6.Write a query to identify which sale_year and category combinations were the most
profitable. Return the columns sale_year, category, the total revenue as total_revenue,
the total cost as total_cost, and the difference between the two as profit for each
sale_year and category. Include only rows that were profitable and sort the results from
highest profit to lowest.

In [38]:
query6_customer = (
    customer_spending
    .groupby(["sale_year", "category"])
    .agg(
        total_revenue=("revenue", "sum"),
        total_cost=("cost", "sum")
    )
    .reset_index()
)

query6_customer["profit"] = (
    query6_customer["total_revenue"] -
    query6_customer["total_cost"]
)

query6_customer = query6_customer[
    query6_customer["profit"] > 0
]

query6_customer = query6_customer.sort_values(
    "profit",
    ascending=False
)

query6_customer

,sale_year,category,total_revenue,total_cost,profit
3,2016,Accessories,4594897,3559646,1035251
4,2016,Bikes,5722257,5209322,512935
5,2016,Clothing,2079651,1654855,424796
0,2015,Accessories,2825767,2482249,343518
2,2015,Clothing,1357906,1237470,120436


### Interpretation

The results show that five sale year and category combinations were profitable. The highest profit was generated by Accessories in 2016, with a profit of 1,035,251. This was followed by Bikes in 2016, with a profit of 512,935, and Clothing in 2016, with a profit of 424,796. Accessories in 2015 generated a profit of 343,518, while Clothing in 2015 generated 120,436.

Only profitable combinations are shown because combinations with zero or negative profit were excluded according to the query requirement.

7.Write a query that calculates the average spending per age for male customers
(gender), where spending is based on the product of unit_price and quantity. Return the age and the calculated average spending, rounded to two decimal places, as
avg_spending. Organize the results from highest to lowest avg_spending.

In [39]:
query7_customer = (
    customer_spending[
        customer_spending["gender"] == "M"
    ]
    .assign(
        spending=lambda x: x["unit_price"] * x["quantity"]
    )
    .groupby("age")
    .agg(
        avg_spending=("spending", "mean")
    )
    .reset_index()
)

query7_customer["avg_spending"] = (
    query7_customer["avg_spending"].round(2)
)

query7_customer = query7_customer.sort_values(
    "avg_spending",
    ascending=False
)

query7_customer

,age,avg_spending
65,86,2026.00
56,73,1170.00
57,74,971.00
58,75,869.00
55,72,764.00
...,...,...
60,78,257.75
61,79,182.60
59,76,159.00
62,81,93.00


### Interpretation

The results show that average spending varies substantially across ages among male customers. The highest average spending is observed for customers aged 86, at 2,026.00. Other relatively high averages occur among customers aged 73, 74, and 75. Because the results are ordered from highest to lowest average spending, the table makes it easy to identify the age groups with the largest average spending values.

8.Write a query to determine the highest unit_cost, lowest unit_cost, and average
unit_cost for each gender in each category. The output columns should be gender,
category, high_cost, low_cost, avg_cost, in that order. Organize the results by gender
and then category.

In [40]:
query8_customer = (
    customer_spending
    .groupby(["gender", "category"])
    .agg(
        high_cost=("unit_cost", "max"),
        low_cost=("unit_cost", "min"),
        avg_cost=("unit_cost", "mean")
    )
    .reset_index()
)

query8_customer["high_cost"] = query8_customer["high_cost"].round(2)
query8_customer["low_cost"] = query8_customer["low_cost"].round(2)
query8_customer["avg_cost"] = query8_customer["avg_cost"].round(2)

query8_customer = query8_customer[
    ["gender", "category", "high_cost", "low_cost", "avg_cost"]
].sort_values(
    ["gender", "category"]
)

query8_customer

,gender,category,high_cost,low_cost,avg_cost
0,F,Accessories,3120.0,0.67,159.30
1,F,Bikes,2443.0,180.00,956.65
2,F,Clothing,2100.0,3.00,334.90
3,M,Accessories,3240.0,0.67,166.96
4,M,Bikes,2443.0,180.00,949.31
5,M,Clothing,2100.0,3.00,337.60


### Interpretation

The results show differences in unit costs across product categories and gender groups. Bikes have the highest average unit cost for both female and male customers, at 956.65 and 949.31 respectively. Accessories have much lower average unit costs, while Clothing falls between the two categories. The highest observed unit cost is 3,240.00 for male customers in Accessories, compared with 3,120.00 for female customers in the same category.

9.Write a query to determine the age distribution of customers by category and country
for the sale_year 2016. For each category and country, calculate the age of the
youngest and oldest customers, and the average age of customers rounded to one
decimal place. The output columns should be category, country, youngest_customer,
oldest_customer, and avg_customer_age, in that order. Organize your results by
category and then by the average age of customers.

In [41]:
query9_customer = (
    customer_spending[
        customer_spending["sale_year"] == 2016
    ]
    .groupby(["category", "country"])
    .agg(
        youngest_customer=("age", "min"),
        oldest_customer=("age", "max"),
        avg_customer_age=("age", "mean")
    )
    .reset_index()
)

query9_customer["avg_customer_age"] = (
    query9_customer["avg_customer_age"].round(1)
)

query9_customer = query9_customer.sort_values(
    ["category", "avg_customer_age"]
)

query9_customer

,category,country,youngest_customer,oldest_customer,avg_customer_age
1,Accessories,Germany,17,87,35.4
0,Accessories,France,17,84,35.6
2,Accessories,United Kingdom,17,85,36.1
3,Accessories,United States,17,78,37.8
4,Bikes,France,17,59,34.0
5,Bikes,Germany,17,72,34.6
6,Bikes,United Kingdom,17,75,34.8
7,Bikes,United States,17,72,39.7
9,Clothing,Germany,17,86,34.2
8,Clothing,France,17,84,35.3


### Interpretation

The 2016 results show differences in customer age distributions across product categories and countries. The youngest customer in every category-country combination was age 17. Average customer age varied across the combinations, with Bikes in the United States having the highest average age at 39.7 years. The results also show that the United States generally had higher average customer ages than the other countries within each product category.

10.Write a query to return the country that has the highest average revenue (rounded to 2
decimals). Your output columns should be country and high_sales.

In [42]:
query10_customer = (
    customer_spending
    .groupby("country")
    .agg(
        high_sales=("revenue", "mean")
    )
    .reset_index()
)

query10_customer["high_sales"] = (
    query10_customer["high_sales"].round(2)
)

query10_customer = query10_customer.sort_values(
    "high_sales",
    ascending=False
).head(1)

query10_customer

,country,high_sales
1,Germany,816.09


### Interpretation

Germany has the highest average revenue among the countries in the dataset, with an average revenue of 816.09 per transaction record. This result identifies Germany as the country with the highest mean revenue based on the available customer spending data.